File: 53. Maximum Subarray.java
class Solution {
    public int maxSubArray(int[] nums) {

        int max = nums[0];
        int msf = 0;

        for(int num : nums){
            max = Math.max(msf + num, max);
            if(num + msf < 0){
                msf = 0;
            } else {
                msf = msf + num;
            }
        }

        return max;
        
    }
}

File: 53. Maximum Subarray.cpp
//Time: O(n)
//Space: O(n)

//Approach: Kadane's Algorithm

class Solution {
public:    
    int maxSubArray(vector<int>& nums) {
        int maxi = -1e9;
        int sum = 0;
        
        for(int u: nums){
            sum += u;
            maxi = max(sum,maxi);
            sum= max(sum,0);
        }
        
        return maxi;
    }
};


